Added new performance test for node buckets

bhroam opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Bug/feature Description
Added new automated performance test for node buckets

#### Solution Description
I wrote a test which compares place=free jobs using both the node bucket algorithm and the normal node search algorithm.

#### Testing logs/output
[node_bucket.log](https://github.com/PBSPro/pbspro/files/2743133/node_bucket.log)


#### Checklist:
<!--- Use the preview button to see the checkboxes/links properly. -->
<!--- Go over the following points, and put an `x` (without spaces around it) in the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [X] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [X] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
   - [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [X] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
   - [ ] I have added  **manual test(s) to this pull request and explained why PTL is not appropriate** for this case.
- [X] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [X] I have attached **test logs to this pull request** as evidence of testing/verification.


__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

agrawalravi90 commented on test/tests/performance/pbs_node_buckets_perf.py:
```diff
@@ -65,13 +65,43 @@ def cust_attr_func(self, name, totalnodes, numnode, attribs):
+        """
+        Run a cycle and return the length of the cycle
+        """
+        t = int(time.time())
```
If the scheduler was already in a scheduling cycle when you executed this then there could be a race condition in this function. I think we should first check that server/scheduler state is idle before doing anything in this function.

bhroam replied:
Good point.  When I added the test, we were in cycle.  I hadn't turned scheduling off before submitting all 1000 jobs.  I did that and checked for server_stat = Idle.

agrawalravi90 commented on test/tests/performance/pbs_node_buckets_perf.py:
```diff
@@ -65,13 +65,43 @@ def cust_attr_func(self, name, totalnodes, numnode, attribs):
+        This test uses place=scatter.  Scatter placement is quicker than free
         """
         num_jobs = 1000
         a = {'Resource_List.select': '1429:ncpus=1:color=yellow',
```
I know this is old code, but can you please explain why you are submitting this job before proceeding with submitting the large number of jobs?

bhroam replied:
It's to make sure the scheduler can't run any of the 1000 jobs.  We want to treat each of the jobs equally, so that means we need to either run them all or not run any.  Submitting one first means we won't run any.  I added a code comment.

bhroam commented:
@agrawalravi90 I updated the tests and cleaned them up a bit.  Since the two tests practically the same, I factored out the entire test and put it into a helper function.  Now the two tests call the helper function with a place statement and the number of jobs.  I modified run_cycle() as you suggested.  I make sure we're not in a cycle.  I kick a cycle and make sure we're in a cycle, and then I wait for that cycle to end.  It's now all done via the server_state attribute instead of parsing logs.  Let me know what you think.

agrawalravi90 commented:
Codacy didn't pass, it's complaining about trailing whitespaces, please fix them.

visheshh commented on test/tests/performance/pbs_node_buckets_perf.py:
```diff
@@ -66,69 +69,97 @@ def cust_attr_func(self, name, totalnodes, numnode, attribs):
         self.logger.info('Cycle 1: %d Cycle 2: %d Cycle time difference: %d' %
                          (cycle1_time, cycle2_time, cycle1_time - cycle2_time))
-        self.assertTrue(cycle1_time > cycle2_time)
+        self.assertGreater(cycle1_time, cycle2_time)
```
Hi @bhroam , Can you make use of set_test_measurements API so that the test results are stored in the JSON output file .
self.set_test_measurements({"test_measure": cycle_time_without_node_buckets,"value": cycle_time})
self.set_test_measurements({"test_measure": cycle_time_with_node_buckets,"value": cycle2_time}).
We are planning to store all results from all the performance tests to json file , So we can use them to compare performance across releases.

agrawalravi90 commented on test/fw/ptl/utils/pbs_logutils.py:
```diff
@@ -920,9 +919,9 @@ class PBSServerLog(PBSLogAnalyzer):
     server_run_tag = re.compile(tm_re + ".*" + job_re + ".*;Job Run at.*")
     server_nodeup_tag = re.compile(tm_re + ".*Node;.*;node up.*")
-    server_enquejob_tag = re.compile(tm_re + ".*" + job_re +
+    server_enquejob_tag = re.compile(tm_re + ".*" + job_re + 
```
looks like your editor configured to add spaces after  '+'

bhroam commented:
@agrawalravi90 Pydev changed its default python formatter.  It was being run when I saved and added the spaces(weird).  I changed it back to running autopep8.  You should make sure it didn't change for you as well.
